sdk/src/main/scala/com/horizen/account/api/rpc/service/EthService.scala

```diff
@@ -348,32 +386,123 @@ class EthService(val scNodeViewHolderRef: ActorRef, val nvtimeout: FiniteDuratio
+  ) = {
     implicit val timeout: Timeout = 5 seconds
     val barrier = Await
       .result(sidechainTransactionActorRef ? BroadcastTransaction(transaction), timeout.duration)
       .asInstanceOf[Future[Unit]]
     onComplete(barrier) {
       // TODO: add correct responses
       case Success(_) => ApiResponseUtil.toResponse(transactionResponseRepresentation(transaction))
-      case Failure(exp) => ApiResponseUtil.toResponse(GenericTransactionError(
-        "GenericTransactionError",
-        JOptional.of(exp)))
+      case Failure(exp) =>
+        ApiResponseUtil.toResponse(GenericTransactionError("GenericTransactionError", JOptional.of(exp)))
     }
   }
 
   @RpcMethod("eth_getCode")
   @RpcOptionalParameters(1)
   def getCode(address: Address, tag: String): String = {
     applyOnAccountView { nodeView =>
-      getStateViewAtTag(nodeView,tag) { tagStateView =>
-          val code = tagStateView.getCode(address.toBytes)
-          if (code == null)
-            "0x"
-          else
-            Numeric.toHexString(code)
+      getStateViewAtTag(nodeView, tag) { tagStateView =>
+        val code = tagStateView.getCode(address.toBytes)
+        if (code == null)
+          "0x"
+        else
+          Numeric.toHexString(code)
+      }
+    }
+  }
+
+  @RpcMethod("debug_traceBlockByNumber")
+  def traceBlockByNumber(blockNumber: String): DebugTraceBlockByIdView = {
+    val currentBlockNumber = Numeric.cleanHexPrefix(blockNumber)
+    val previousBlockNumber = Numeric.cleanHexPrefix((java.lang.Long.decode(currentBlockNumber) - 1).toHexString)
+
+    applyOnAccountView { nodeView =>
+      getStateViewAtTag(nodeView, previousBlockNumber) { tagStateView =>
+        {
+          val requestedBlockId = getBlockIdByTag(nodeView, currentBlockNumber)
+          val requestedBlock = nodeView.history.getBlockById(requestedBlockId).get()
+          val transactions = requestedBlock.transactions
+
+          for (mcBlockRefData <- requestedBlock.mainchainBlockReferencesData) {
+            tagStateView.applyMainchainBlockReferenceData(mcBlockRefData).get
+          }
+
+          val evmResults = transactions.map(tx => {
+            Evm.Trace(
+              tagStateView.getStateDbHandle,
+              if (tx.getFrom == null) null
+              else tx.getFrom.bytes(),
+              if (tx.getTo == null) null
+              else tx.getTo.bytes(),
+              tx.getValue,
+              tx.getData,
+              tx.getGasLimit,
+              tx.getGasPrice,
+              null
+            )
+          })
+
+          new DebugTraceBlockByIdView(evmResults.toArray)
+        }
+      }
+    }
+  }
+
+  @RpcMethod("debug_traceTransaction")
+  def traceTransaction(transactionHash: String): DebugTraceTransactionView = {
+    val requestedTransaction = getTransactionAndReceipt(transactionHash) { (tx, receipt) =>
+      new EthereumTransactionView(receipt, tx)
+    }.orNull
```

Comment:
in case transaction not found we need to return proper result message.
Otherwise we will throw an null pointer exception.
Let's add the case with unknown tx id in the python test.

Reply:
still actual